agent/functional_tests/util/utils.go

```diff
@@ -886,3 +887,60 @@ func GetTaskID(taskARN string) (string, error) {
 
 	return resourceSplit[1], nil
 }
+
+// WaitContainerInstanceStatus waits for a container instance to reach certain status by polling its status
+func (agent *TestAgent) WaitContainerInstanceStatus(desiredStatus string) error {
+	timer := time.NewTimer(waitContainerInstanceStatusDuration)
+	errChan := make(chan error, 1)
+	containerInstanceStatus := ""
+
+	cancelled := false
+	go func() {
+		for !cancelled {
+			status, err := agent.getContainerInstanceStatus()
+			if err != nil {
+				errChan <- err
+				return
+			}
+			containerInstanceStatus = status
+
+			if status == desiredStatus {
+				break
+			}
+			if desiredStatus == "ACTIVE" {
+				if status == "REGISTRATION_FAILED" || status == "INACTIVE" {
+					errChan <- errors.Errorf("container instance ends at status %s; will never reach ACTIVE", status)
+					return
+				}
+			}
+			time.Sleep(5 * time.Second)
+		}
+		errChan <- nil
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-timer.C:
+		cancelled = true
+		return errors.Errorf("timed out waiting for container instance '%s' to reach 'ACTIVE', status is '%s'",
+			agent.ContainerInstanceArn, containerInstanceStatus)
```

Comment:
What happens if the desiredStatus is "INACTIVE" but something goes wrong? Any error message or something else to indicate that the trunk is not detached successfully?

Reply:
here i'm assuming the trunk is detached when instance goes to INACTIVE, which is the expected behavior. I'm adding a test [here](https://github.com/aws/amazon-ecs-agent/pull/1961) to check whether that's true. i don't think we need to check this in many tests, since if somehow the logic breaks there's already a failing test that can inform us of the issue.

Reply:
maybe i misunderstood your concern. if backend failed to detach the trunk, the instance will still go to INACTIVE but not very soon. we will time out here.

Reply:
Then in that case, we might need to output an error message like "timed out waiting for container instance to reach "INACTIVE" because of the failure of trunk detachment" separately?

Reply:
in case of timeout, the agent can't tell whether backend has tried something and failed, or it's slow and hasn't tried anything.